web/src/components/storage/DevicesManager.js

```diff
@@ -0,0 +1,207 @@
+   * @param {StorageDevice[]} staging - Devices representing the target state of the system.
+   * @param {Action[]} actions - Actions to perform from system to staging.
+   */
+  constructor(system, staging, actions) {
+    this.system = system;
+    this.staging = staging;
+    this.actions = actions;
+  }
+
+  /**
+   * System device with the given SID.
+   *
+   * @param {Number} sid
+   * @returns {StorageDevice|undefined}
+   */
+  systemDevice(sid) {
+    return this.#device(sid, this.system);
+  }
+
+  /**
+   * Staging device with the given SID.
+   *
+   * @param {Number} sid
+   * @returns {StorageDevice|undefined}
+   */
+  stagingDevice(sid) {
+    return this.#device(sid, this.staging);
+  }
+
+  /**
+   * Whether the given device exists in system.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  existInSystem(device) {
+    return this.#exist(device, this.system);
+  }
+
+  /**
+   * Whether the given device exists in staging.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  existInStaging(device) {
+    return this.#exist(device, this.staging);
+  }
+
+  /**
+   * Whether the given device is going to be formatted.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  hasNewFilesystem(device) {
+    if (!device.filesystem) return false;
+
+    const systemDevice = this.systemDevice(device.sid);
+    const systemFilesystemSID = systemDevice?.filesystem?.sid;
+
+    return device.filesystem.sid !== systemFilesystemSID;
+  }
+
+  /**
+   * Whether the given device is going to be shrunk.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  isShrunk(device) {
+    return this.shrinkSize(device) > 0;
+  }
+
+  /**
+   * Amount of bytes the given device is going to be shrunk.
+   *
+   * @param {StorageDevice} device
+   * @returns {Number}
+   */
+  shrinkSize(device) {
+    const systemDevice = this.systemDevice(device.sid);
+    const stagingDevice = this.stagingDevice(device.sid);
+
+    if (!systemDevice || !stagingDevice) return 0;
+
+    const amount = systemDevice.size - stagingDevice.size;
+    return amount > 0 ? amount : 0;
+  }
+
+  /**
+   * Disk devices and LVM volume groups used for the installation.
+   *
+   * @note The used devices are extracted from the actions.
+   *
+   * @returns {StorageDevice[]}
+   */
+  usedDevices() {
+    const isTarget = (device) => device.isDrive || device.type === "lvmVg";
+
+    // Check in system devices to detect removals.
+    const targetSystem = this.system.filter(isTarget);
+    const targetStaging = this.staging.filter(isTarget);
+
+    const sids = targetSystem.concat(targetStaging)
+      .filter(d => this.#isUsed(d))
+      .map(d => d.sid);
+
+    return compact(uniq(sids).map(sid => this.stagingDevice(sid)));
+  }
+
+  /**
+   * Devices deleted.
+   *
+   * @note The devices are extracted from the actions.
+   *
+   * @returns {StorageDevice[]}
+   */
+  deletedDevices() {
```

Comment:
I wonder if that method could receive a flag for excluding subvolumes. Use case when rendering the amount of planned delete actions. What is more, I wonder why we have to filter the subvolume actions there.

Reply:
> I wonder if that method could receive a flag for excluding subvolumes. Use case when rendering the amount of planned delete actions. 

This method cannot be used for counting the amount of delete actions. It returns the deleted devices (e.g., partitions, VGs, MD RAIDs, etc). The number of delete actions can be more than the number of deleted devices.

To know the number of delete actions:

`devicesManager.actions.filter(a => a.delete).length`

And if you want to exclude the subvolume actions, then:

`devicesManager.actions.filter(a => a.delete && !a.subvol).length`

> What is more, I wonder why we have to filter the subvolume actions there.

I don't get this. Where are the subvolume actions filtered out?

Reply:
> > What is more, I wonder why we have to filter the subvolume actions there.
> 
> I don't get this. Where are the subvolume actions filtered out?

Nevermind. I meant that we have a method for filtering the subvolumes from actions in the ProposalResultSection.jsx at line ~232. But we have a "DevicesManager" object that provides a "deleteDevices" method and, in my opinion, could provide a "deleteSystems" too. Although I konw, it is a "DeviceManager" not a "SystemsManager".